bazel_utils.go

```
@@ -0,0 +1,182 @@
+		var err error
+		f, err = os.CreateTemp("", "gcassert-object-code-inspection*.log")
+		if err != nil {
+			return err
+		}
+		mw = io.MultiWriter(pw, f)
+		defer fmt.Printf("See %s for full output.\n", f.Name())
+	}
+
+	disassembleCmdError := make(chan error, 1)
+	go func() {
+		defer func() {
+			_ = pw.Close()
+			if f != nil {
+				_ = f.Close()
+			}
+		}()
+		for _, path := range paths {
+			getPackageName := func(path string) string {
+				return path[strings.LastIndex(path, "/")+1:]
+			}
+
+			absPath, err := filepath.Abs(path)
+			if err != nil {
+				disassembleCmdError <- err
+				return
+			}
+			relPkgPath, err := filepath.Rel(workspacePath, absPath)
+			if err != nil {
+				disassembleCmdError <- err
+				return
+			}
+			objectFilePath := filepath.Join(bazelBin, relPkgPath, getPackageName(path)+".a")
+			args := []string{"tool", "objdump", objectFilePath}
+			objectFileDisassembleCmd := exec.Command("go", args...)
+			if debugMode {
+				objectFileDisassembleCmd.Stdout = mw
+				objectFileDisassembleCmd.Stderr = mw
+				// Log full 'objdump' command.
+				fmt.Fprintln(f, objectFileDisassembleCmd)
+			} else {
+				objectFileDisassembleCmd.Stdout = pw
+				objectFileDisassembleCmd.Stderr = pw
+			}
+
+			if err := objectFileDisassembleCmd.Run(); err != nil {
+				disassembleCmdError <- err
+				return
+			}
+		}
+		disassembleCmdError <- nil
+	}()
+
+	scanner := bufio.NewScanner(pr)
+
+	// callRegex is used to read a function identifier from a CALL line in the object file.
+	// Example call line from the object file:
+	/*
+	  mvcc.go:1135			0x66928a		94000000		CALL 0(PC)									[0:4]R_CALLARM64:github.com/cockroachdb/cockroach/pkg/storage.(*MVCCGetOptions).validate
+
+	*/
+	callRegex, err := regexp.Compile(`\s*(\S+)\s*\S*\s*\S*\s*(CALL)\s*\S*\s*(\S+)`)
+	if err != nil {
+		return err
+	}
+	for scanner.Scan() {
+		line := scanner.Text()
+		fnIdentifierMatch := callRegex.FindStringSubmatch(line)
+		if fnIdentifierMatch != nil {
```

To, at least partially, address my concern about silently doing nothing, perhaps we could add an assertion that `callRegex` matched at least once and return an error if not (hidden behind a new bool flag of `gcassert` that we would enable in our CI since we do expect some non-inlined functions in all packages in cockroachdb code base on which we'll run `gcassert`). Thoughts?

---

We can't use a flag because we will run gcassert using the library not the binary. I did something close. Let me know what you think.

---

I like it, thanks!